db/versions/0017.yml

```diff
@@ -0,0 +1,699 @@
+          indexed_tasks.namespace,
+          indexed_tasks.name,
+          indexed_tasks.rank,
+          indexed_tasks.task_id,
+          indexed_tasks.data,
+          indexed_tasks.expires
+        from indexed_tasks
+        where
+          (indexed_tasks.namespace = namespace_in or namespace_in is null) and
+          (indexed_tasks.name = name_in or name_in is null) and
+          (indexed_tasks.expires > now())
+        -- we previously used to order by the hashed namespace but there's probably no need to add this complication.
+        order by indexed_tasks.namespace, name
+        limit get_page_limit(page_size_in)
+        offset get_page_offset(page_offset_in);
+      end
+  create_indexed_task:
+    description: |-
+      Create a new indexed task. Raises UNIQUE_VIOLATION if the indexed task already exists.
+      Returns the etag of the newly created indexed task.
+    mode: write
+    serviceName: index
+    args: namespace_in text, name_in text, rank_in integer, task_id_in text, data_in jsonb, expires_in timestamptz
+    returns: uuid
+    body: |-
+      declare
+        new_etag uuid := public.gen_random_uuid();
+      begin
+        insert
+          into indexed_tasks (namespace, name, rank, task_id, data, expires, etag)
+          values (namespace_in, name_in, rank_in, task_id_in, data_in, expires_in, new_etag);
+        return new_etag;
+      end
+  get_indexed_task:
+    description: |-
+      Get a non-expired indexed task. The returned table will have one or
+      (if no such indexed task is defined) an exception will be raised with error code 'P0002'.
+    mode: read
+    serviceName: index
+    args: namespace_in text, name_in text
+    returns: table(namespace text, name text, rank integer, task_id text, data jsonb, expires timestamptz, etag uuid)
+    body: |-
+      declare
+        new_row indexed_tasks%ROWTYPE;
+      begin
+        select
+          indexed_tasks.namespace,
+          indexed_tasks.name,
+          indexed_tasks.rank,
+          indexed_tasks.task_id,
+          indexed_tasks.data,
+          indexed_tasks.expires,
+          indexed_tasks.etag
+        from indexed_tasks
+        where
+          indexed_tasks.namespace = namespace_in and
+          indexed_tasks.name = name_in and
+          indexed_tasks.expires > now()
+        into new_row;
+        if found then
+          return query select new_row.namespace, new_row.name, new_row.rank, new_row.task_id, new_row.data, new_row.expires, new_row.etag;
+          return;
+        end if;
+
+        raise exception 'no such row' using errcode = 'P0002';
+      end
+  update_indexed_task:
+    serviceName: index
+    description: |-
+      Update an indexed task.
+      Returns the up-to-date indexed task row that have the same namespace and name.
+      If the etag argument is empty then the update will overwrite the matched row.
+      Else, the function will fail if the etag is out of date. This is useful for concurency handling.
```

Comment:
I only see this called in one place, and it seems that the JS doesn't catch this error.  Is the eventual consistency necessary here?

Reply:
Probably not. I will remove it.